geocruncher/profiler.py

```diff
@@ -60,71 +72,107 @@ class ProfilerSettings(NamedTuple):
-_metadata = dict()
-_last_profiled = None
-
-
-def init_profiler(settings: ProfilerSettings):
-    """Init the profiler. To be called once at the beginning of the geocruncher process"""
-    global _settings, _steps, _last_profiled, _metadata
-    # set the start time, to calculate relative durations
-    _last_profiled = time.process_time()
-    _settings = settings
-    # make the dictionnary with a default value for each step
-    _steps = dict([[step, 0] for step in _settings.steps])
-    # make the dictionnary with a default value for each metadata
-    _metadata = dict([[meta, None] for meta in _settings.metadata])
-
-    # set the metadata that's on every computation
-    set_profiler_metadata("start_time", datetime.datetime.utcnow().isoformat())
-
-
-def profile(step: str):
-    """Profile a step by name. To be called when the step in question is done.
-    Also works inside loops (total time per step gets summed up). For loops, it is recommanded to also profile just before entering, and at the end of every iteration, so the first profile in the loop is consistant"""
-    global _steps, _last_profiled
-    if _settings is not None and step in _steps:
-        now = time.process_time()
-        # add for the step the difference between now and the last profiled time
-        # since we add, the profile function works in loops. on each loop iteration, the step's time will be increased
-        _steps[step] += now - _last_profiled
-        _last_profiled = now
-
-
-def set_profiler_metadata(metadata: str, value):
-    global _metadata
-    if metadata in _metadata:
-        _metadata[metadata] = value
-
-
-def save_profiler_results():
-    """For now, make a JSON file for each profiling"""
-    global _steps, _settings
-    if _settings is not None:
+# Profiler class
+
+
+class VkProfiler():
+    def __init__(self, settings: VkProfilerSettings):
+
+        # set the start time, to calculate relative durations
+        self._last_profiled = time.process_time()
+        self._settings = settings
+        # dictionnary where each step's name maps to the total fractional seconds spent on that step
+        # make the dictionnary with a default value for each step
+        self._steps = dict(
+            [[step, {'profile': list(), 'time': 0}] for step in settings.steps])
+        # make the dictionnary with a default value for each metadata
+        self._metadata = dict([[meta, None] for meta in settings.metadata])
+
+        # set the metadata that's on every computation
+        self.set_profiler_metadata(
+            'start_time', datetime.datetime.utcnow().isoformat())
+        pr.enable()
+
+    def profile(self, step: str):
+        """Profile a step by name. To be called when the step in question is done.
+        Also works inside loops (total time per step gets summed up). For loops, it is recommanded to profile just before entering, and at the end of every iteration, so the first profile in the loop is consistant"""
```

Comment:
No, it is recommended to profile "any" step before entering a loop, so that the delta time in the first loop iteration will closely match the delta time of the next iterations, assuming we also profile at the end of every loop iteration.

I don't think it's worth having a dummy step or anything, as it makes it more complex, and in the end probably makes 1/10 - 1 milliseconds difference when we will be profiling minutes - hours.

Because of this, there are time when I did not profile before a loop. I believe I left a comment when I did that.